stackslib/src/chainstate/nakamoto/shadow.rs

```diff
@@ -0,0 +1,854 @@
+    /// expire.
+    ///
+    /// It's used to create shadow blocks.
+    pub(crate) fn shadow_load_tenure_info<'a>(
+        &self,
+        chainstate: &'a mut StacksChainState,
+        burn_dbconn: &'a SortitionHandleConn,
+        cause: Option<TenureChangeCause>,
+    ) -> Result<MinerTenureInfo<'a>, Error> {
+        self.inner_load_tenure_info(chainstate, burn_dbconn, cause, true)
+    }
+
+    /// Begin/resume mining a shadow tenure's transactions.
+    /// Returns an open ClarityTx for mining the block.
+    /// NOTE: even though we don't yet know the block hash, the Clarity VM ensures that a
+    /// transaction can't query information about the _current_ block (i.e. information that is not
+    /// yet known).
+    pub fn shadow_tenure_begin<'a, 'b>(
+        &mut self,
+        burn_dbconn: &'a SortitionHandleConn,
+        info: &'b mut MinerTenureInfo<'a>,
+        tenure_id_consensus_hash: &ConsensusHash,
+    ) -> Result<ClarityTx<'b, 'b>, Error> {
+        let tenure_snapshot = SortitionDB::get_block_snapshot_consensus(
+            burn_dbconn.conn(),
+            tenure_id_consensus_hash,
+        )?
+        .ok_or_else(|| Error::NoSuchBlockError)?;
+
+        let SetupBlockResult {
+            clarity_tx,
+            matured_miner_rewards_opt,
+            ..
+        } = NakamotoChainState::setup_shadow_block_processing(
+            &mut info.chainstate_tx,
+            info.clarity_instance,
+            burn_dbconn,
+            burn_dbconn.context.first_block_height,
+            &burn_dbconn.context.pox_constants,
+            info.parent_consensus_hash,
+            info.parent_header_hash,
+            info.parent_burn_block_height,
+            tenure_snapshot,
+            info.cause == Some(TenureChangeCause::BlockFound),
+            info.coinbase_height,
+            info.cause == Some(TenureChangeCause::Extended),
+        )?;
+        self.matured_miner_rewards_opt = matured_miner_rewards_opt;
+        Ok(clarity_tx)
+    }
+
+    /// Get an address's account
+    fn get_account(
+        chainstate: &mut StacksChainState,
+        sortdb: &SortitionDB,
+        addr: &StacksAddress,
+        tip: &StacksHeaderInfo,
+    ) -> Result<StacksAccount, Error> {
+        let snapshot =
+            SortitionDB::get_block_snapshot_consensus(&sortdb.conn(), &tip.consensus_hash)?
+                .ok_or_else(|| Error::NoSuchBlockError)?;
+
+        let account = chainstate
+            .with_read_only_clarity_tx(
+                &sortdb.index_handle(&snapshot.sortition_id),
+                &tip.index_block_hash(),
+                |clarity_conn| {
+                    StacksChainState::get_account(clarity_conn, &addr.to_account_principal())
+                },
+            )
+            .ok_or_else(|| Error::NoSuchBlockError)?;
+
+        Ok(account)
+    }
+
+    /// Make a shadow block from transactions
+    pub fn make_shadow_block_from_txs(
```

Comment:
Could this and `make_shadow_tenure` be gated to ensure it isn't called by the node? I assume you'll need this for the tooling (haven't started reviewing that PR yet), otherwise it could be gated just for tests.

Reply:
It's needed by the tooling.